utils/loading_shared_values.hh

```diff
@@ -260,6 +260,30 @@ public:
         }
     }
 
+    /// \brief
+    /// Returns an entry_ptr for the given key. If the value is already present, it is returned.
+    /// If the entry is loading, or it is missing, the parameter is used as the value.
+    entry_ptr get_or_insert(const key_type& key, value_type value) {
```

Comment:
The whole idea of this API makes very little sense to me.

I'd like to start from the beginning which is the following statement in the description of the patch:

```
we will want to use the loading_cache
to store udf instances to reuse them. Because the used
instances cannot be used by multiple users simultaneously,
we will want to remove the instance from the cache before
each use, and reinsert the instance to the cache afterwards.
```

The problem of "not letting multiple users use the same value simultaneously" has a standard solution called "a semaphore/mutex". All you need to do is to have a semaphore (mutex) to be part of your "value" and then every user should simply try to get it before actually using the value. 

That's it - no need to change the container.

Let me know if that makes sense?

Reply:
The issue with storing udf instances in the cache is that after each use they may change their size and currently the tracked size in the cache is only adjusted when adding/removing entries.
We could still use this approach if we added some methods for modifying the tracked size to the cache, but when attempting to implement that it turned out more complicated than the posted solution.
I had a line about the mutable size of the instances in the first version, but I replaced that with the mention of simultaneous use of the instances. Maybe I should have kept both

Reply:
UDFs aside; I agree that `get_or_insert` sounds extremely weird as part of public API on its own, but, if I understand correctly, it's just a counterpart for `get_or_load` which allows us to implement a synchronous `loading_cache::put()` method, right? And this API, `put()`, sounds like a very useful addition to the structure - if you're already holding a value and want to put it in the cache without going through the loading mechanism, which involves futures, yielding and allocations, it sounds useful to be able to insert the value in a more efficient way, and also atomically.

Reply:
I still don't understand why do you need this API. 
I also don't think it a good idea to use a cache size in order to limit the maximum memory consumption size by the UDF given what you said above.

`loading_cache` class size semantics is, indeed as you have described, assumes that the size is calculated during the item creation and I don't see how you can possibly implement a "dynamic item size" without breaking the "maximum container size" guarantee.

I think your should be using a separate memory permit for UDF which is unrelated to the `loading_cache` container.

Please, let me know if that makes more sense now, @wmitros ?

Reply:
The size thing is just a distraction. The problem here is that the cached value is mutated by its user. While in the cache all instances of the same key are equivalent, but once out of the cache they are not, each is a snowflake.

A semaphore would prevent mutability being a problem, but would also prevent concurrent execution of the same UDF.

It's similar to AWS Lambda. All idle lambdas of the same function are equivalent, but once an idle lambda starts executing it cannot be used to serve another invocation until it is complete (in fact AWS lambda is similar to a traditional database trigger).

Reply:
> The size thing is just a distraction. 

I agree. That's why I wrote that this PR doesn't seem to address the problem.

> It's similar to AWS Lambda. All idle lambdas of the same function are equivalent, but once an idle lambda starts executing it cannot be used to serve another invocation until it is complete (in fact AWS lambda is similar to a traditional database trigger).
> A semaphore would prevent mutability being a problem, but would also prevent concurrent execution of the same UDF.

I'm confused: in your "AWS lambda" you seem to suggest that a specific UDF execution has to be serialized. But then you say you don't want to prevent a concurrent execution of the same UDF.

If what you mean in your "snowflake" metaphor on top of "concurrent execution of the same UDF" is that some UDF "snoflakes" would still remain the same (and these should be serialized) but some can change and these should not serialize with the former then you should probably include the parts of the UDF state that represent the "difference between snowflakes" to be part of the hash key.

Then, when you detect a "creation of such a new instance" you should go via the API that allows you to provide a loader function on every `get_ptr()` call and this would allow you inserting the "already existing" object into the cache without having to create it: `future<value_ptr> get_ptr(const Key& k, LoadFunc&& load)`.

Does it make sense?

Reply:
As long as UDFs are CPU-bound, which is the case now (and for the foreseeable future?), higher concurrency would just increase average latency, right?

Reply:
> and it sounds reasonable to remove old, unused instances to make space for new, frequently used instances (even before they timeout)

@wmitros `loading_cache` has various APIs to remove entries from the cache based on various criteria. Look for "remove" in the loading_cache.hh"

Reply:
> > I don't see how we can remove the least recently used entry using either of these methods.
> 
> You don't have to. This is the magic of LRU eviction - you simply add a new entry and the least recently used is going to be evicted auto-magically. ;)

The LRU eviction doesn't really work if we're not tracking the sizes of entries in the cache and instead leave that to a separate memory permit (that's why you recommended me to look for "remove" in loading_cache.hh, or was it just a distraction?)

> > Assuming we used get_ptr for inserting ready values, we may discard them and be forced to load the values again.
> 
> Not sure I see how this can technically happen since all your lambda's are not going to preempt hence there will be no chance of remove_if() to hit you in the middle. Am I missing anything?

We insert an unready entry before starting executing the lambda: https://github.com/scylladb/scylla/blob/adda43edc75b901b2329bca8f3eb74596698d05f/utils/loading_shared_values.hh#L234-L238
The preemption point I'm concerned about is the start of line 238.
Unless there is a reason we cannot preempt before executing the lambda in 238, if so I guess I need to revise my knowledge about seastar futures

Reply:
Actually, we're not waiting for the future in line 238 directly, so there is no preemption there, but we're waiting for it indirectly in https://github.com/scylladb/scylla/blob/adda43edc75b901b2329bca8f3eb74596698d05f/utils/loading_shared_values.hh#L249 and that's the line I'm worried about. Looks like I need to read about futures again regardless

Reply:
The line that prevents the race you are worried about is above lines you look at.
You should be looking at the line https://github.com/scylladb/scylla/blob/adda43edc75b901b2329bca8f3eb74596698d05f/utils/loading_shared_values.hh#L222.

All but the first caller of a get_ptr() for a given key is NOT going to invoke the loader but rather is going to wait for shared_future (https://github.com/scylladb/scylla/blob/adda43edc75b901b2329bca8f3eb74596698d05f/utils/loading_shared_values.hh#L229) that is going to be set by the first caller.

@wmitros